include/mbgl/storage/default_file_source.hpp

```diff
@@ -140,10 +141,14 @@ class DefaultFileSource : public FileSource {
     class Impl;
 
 private:
+    // Shared so destruction is done on this thread
```

Comment:
@m12 This makes use of the fact that the destructor of our `util::Thread` class [waits for the thread to finish](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/util/thread.hpp#L169). Given the reverse destruction order of members, the thread with associated data is destroyed before the shared_ptr (`DefaultFileSource::assetFileSource`) is released on the main thread and `~AssetFileSource` is called.